libraries/ms-drivers/test/test_mcp3427_adc.c

```diff
@@ -0,0 +1,371 @@
+  })
+
+static uint16_t s_times_callback_called = 0;
+static void *s_callback_context = NULL;
+static uint16_t s_times_fault_callback_called = 0;
+static void *s_fault_callback_context = NULL;
+
+static void prv_callback(int16_t value_ch1, int16_t value_ch2, void *context) {
+  LOG_DEBUG("Callback called: value_ch1=%d, value_ch2=%d\n", value_ch1, value_ch2);
+  s_times_callback_called++;
+  s_callback_context = context;
+}
+
+static void prv_fault_callback(void *context) {
+  LOG_WARN("Fault callback called!");
+  s_times_fault_callback_called++;
+  s_fault_callback_context = context;
+}
+
+void setup_test(void) {
+  event_queue_init();
+  interrupt_init();
+  soft_timer_init();
+
+  I2CSettings i2c_settings = {
+    .speed = I2C_SPEED_FAST,
+    .sda = TEST_CONFIG_PIN_I2C_SDA,
+    .scl = TEST_CONFIG_PIN_I2C_SCL,
+  };
+  i2c_init(TEST_I2C_PORT, &i2c_settings);
+
+  s_times_callback_called = s_times_fault_callback_called = 0;
+  s_callback_context = s_fault_callback_context = NULL;
+}
+void teardown_test(void) {}
+
+// Test a single data fetching round (2 conversions) in the current MCP3427 settings.
+static void prv_test_data_round(Mcp3427Storage *storage, Event *e, void *callback_context) {
+  // triggering CH1 conversion, a data ready event should be raised in MCP3427_MAX_CONV_TIME_MS
+  TEST_ASSERT_OK(mcp3427_process_event(e));
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  delay_ms(MCP3427_MAX_CONV_TIME_MS);
+  TEST_ASSERT_EVENT_WITH_ID(*e, TEST_MCP3427_DATA_READY_EVENT);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  TEST_ASSERT_EQUAL(0, s_times_callback_called);  // callback not called yet
+
+  // reading CH1 data, a data trigger event should be raised immediately
+  TEST_ASSERT_OK(mcp3427_process_event(e));
+  TEST_ASSERT_EVENT_WITH_ID(*e, TEST_MCP3427_DATA_TRIGGER_EVENT);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  TEST_ASSERT_EQUAL(0, s_times_callback_called);  // callback not called yet, only CH1 read
+
+  // triggering CH2 conversion, data ready event raised in MCP3427_MAX_CONV_TIME_MS
+  TEST_ASSERT_OK(mcp3427_process_event(e));
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  delay_ms(MCP3427_MAX_CONV_TIME_MS);
+  TEST_ASSERT_EVENT_WITH_ID(*e, TEST_MCP3427_DATA_READY_EVENT);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  TEST_ASSERT_EQUAL(0, s_times_callback_called);  // callback not called yet
+
+  // reading CH2 data, callback called and data trigger event raised
+  TEST_ASSERT_OK(mcp3427_process_event(e));
+  TEST_ASSERT_EVENT_WITH_ID(*e, TEST_MCP3427_DATA_TRIGGER_EVENT);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);            // callback called since it's CH2
+  TEST_ASSERT_EQUAL(callback_context, s_callback_context);  // context was passed
+
+  // the fault callback shouldn't have been called ever
+  TEST_ASSERT_EQUAL(0, s_times_fault_callback_called);
+  TEST_ASSERT_EQUAL(NULL, s_fault_callback_context);
+
+  // clean up for a possible next round
+  s_callback_context = NULL;
```

Comment:
you skipped this in your test for the sense module since it's in setup_test. Could also put it in teardown_test, but I don't think this should go here.